src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -80,18 +78,188 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
 
             var semanticVersion = await GetDocumentSemanticVersionAsync(documentSnapshot).ConfigureAwait(false);
 
-            // If we have a matching cached response, avoid computation and return early.
-            if (_cachedResponses.TryGetValue(textDocumentIdentifier.Uri, out var response) &&
-                response.IsCSharpFinalized &&
-                response.SemanticVersion == semanticVersion &&
-                response.Range == range)
+            // See if we can use our cache to at least partially avoid recomputation.
+            if (!_tokensCache.TryGetCachedTokens(textDocumentIdentifier.Uri, semanticVersion, range, out var cachedResult))
             {
-                return response.SemanticTokens;
+                // No cache results found, so we'll recompute tokens for the entire range and then hopefully cache the
+                // results to use next time around.
+                var (tokens, isCSharpFinalized) = await GetSemanticTokensAsync(
+                    textDocumentIdentifier, documentSnapshot, documentVersion, range, cancellationToken);
+                if (isCSharpFinalized && tokens is not null)
+                {
+                    _tokensCache.CacheTokens(textDocumentIdentifier.Uri, semanticVersion, range, tokens.Data.ToArray());
+                }
+
+                return tokens;
+            }
+
+            var cachedRange = cachedResult.Value.Range;
+            var cachedTokens = cachedResult.Value.Tokens;
+
+            // The cache returned to us an exact match for the range we're looking for, so we can return early.
+            if (cachedRange.Equals(range))
+            {
+                return new SemanticTokens { Data = cachedTokens.ToImmutableArray() };
             }
 
-            var tokens = await GetSemanticTokensAsync(
-                textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion, range, cancellationToken);
-            return tokens;
+            // The cache returned to us only a partial match for the range. We'll need to compute the rest by
+            // sending range requests to the Razor/C#/HTML servers.
+            var filledInRange = await FillInRangeGapsAsync(
+                textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion,
+                range, cachedRange, cachedTokens, cancellationToken).ConfigureAwait(false);
+
+            return filledInRange;
+        }
+
+        private async Task<SemanticTokens?> FillInRangeGapsAsync(
+            TextDocumentIdentifier textDocumentIdentifier,
+            DocumentSnapshot documentSnapshot,
+            int documentVersion,
+            VersionStamp semanticVersion,
+            Range range,
+            Range cachedRange,
+            List<int> cachedTokens,
+            CancellationToken cancellationToken)
+        {
+            var tokenList = new List<int>();
+            var absoluteLine = 0;
+
+            // There are 3 scenarios:
```

Comment:
Your code seems to account for it but maybe call out in this comment that these scenarios are not mutually exclusive and may contain multiple gaps: IE (5-20, 25-30, 40-50), and a range request comes for 2-60.

Reply:
lol so, I think I made the numbering a little confusing. Technically these scenarios are mutually exclusive in that we only retrieve one cached result and it can either be at the start, middle, or end of the range (or it can be a complete match, but in that case we would have returned before this point).

I updated the wording/numbering a bit, hopefully that helps clarify it?